source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc

```diff
@@ -0,0 +1,322 @@
+#include "source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.h"
+
+#include "source/common/common/regex.h"
+#include "source/common/network/utility.h"
+#include "source/extensions/filters/network/thrift_proxy/auto_protocol_impl.h"
+#include "source/extensions/filters/network/thrift_proxy/auto_transport_impl.h"
+#include "source/extensions/filters/network/thrift_proxy/binary_protocol_impl.h"
+#include "source/extensions/filters/network/thrift_proxy/compact_protocol_impl.h"
+#include "source/extensions/filters/network/thrift_proxy/framed_transport_impl.h"
+#include "source/extensions/filters/network/thrift_proxy/header_transport_impl.h"
+#include "source/extensions/filters/network/thrift_proxy/unframed_transport_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ThriftFilters {
+namespace PayloadToMetadataFilter {
+
+using namespace Envoy::Extensions::NetworkFilters;
+using FieldSelector = envoy::extensions::filters::network::thrift_proxy::filters::
+    payload_to_metadata::v3::PayloadToMetadata::FieldSelector;
+
+Config::Config(const envoy::extensions::filters::network::thrift_proxy::filters::
+                   payload_to_metadata::v3::PayloadToMetadata& config) {
+  trie_root_ = std::make_shared<Trie>();
+  request_rules_.reserve(config.request_rules().size());
+  for (const auto& entry : config.request_rules()) {
+    request_rules_.emplace_back(entry, static_cast<uint16_t>(request_rules_.size()), trie_root_);
+  }
+}
+
+Rule::Rule(const ProtoRule& rule, uint16_t rule_id, TrieSharedPtr root)
+    : rule_(rule), rule_id_(rule_id) {
+  if (!rule_.has_on_present() && !rule_.has_on_missing()) {
+    throw EnvoyException("payload to metadata filter: neither `on_present` nor `on_missing` set");
+  }
+
+  if (rule_.has_on_missing() && rule_.on_missing().value().empty()) {
+    throw EnvoyException(
+        "payload to metadata filter: cannot specify on_missing rule without non-empty value");
+  }
+
+  if (rule_.has_on_present() && rule_.on_present().has_regex_value_rewrite()) {
+    const auto& rewrite_spec = rule_.on_present().regex_value_rewrite();
+    regex_rewrite_ = Regex::Utility::parseRegex(rewrite_spec.pattern());
+    regex_rewrite_substitution_ = rewrite_spec.substitution();
+  }
+
+  switch (rule_.match_specifier_case()) {
+  case ProtoRule::MatchSpecifierCase::kMethodName:
+    match_type_ = MatchType::MethodName;
+    method_or_service_name_ = rule_.method_name();
+    break;
+  case ProtoRule::MatchSpecifierCase::kServiceName:
+    match_type_ = MatchType::ServiceName;
+    if (!rule_.service_name().empty() && !absl::EndsWith(rule_.service_name(), ":")) {
+      method_or_service_name_ = rule_.service_name() + ":";
+    } else {
+      method_or_service_name_ = rule_.service_name();
+    }
+    break;
+  case ProtoRule::MatchSpecifierCase::MATCH_SPECIFIER_NOT_SET:
+    PANIC_DUE_TO_CORRUPT_ENUM;
+  }
+
+  FieldSelector field_selector = rule_.field_selector();
```

Comment:
minor: I think we can make this a `FieldSelector*` or `FieldSelector&` and avoid Swap and the need to explain it later.

Reply:
pointer works. thanks for great suggestion